src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,281 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Fox Thomson.
+-/
+import data.fintype.basic
+import data.finset.basic
+import tactic.rcases
+import tactic.omega
+import computability.language
+
+/-!
+# Regular Expressions
+
+This file contains the formal definition for regular expressions and basic lemmas. Note these are
+regular expressions in terms of formal language theory. Note this is different to regex's used in
+computer science such as the POSIX standard.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/--
+This is the definition of regular expressions. The names used here is to mirror the definition
+of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+* `0` (`zero`) matches nothing
+* `1` (`epsilon`) matches only the empty string
+* `char a` matches only the string 'a'
+* `star P` matches any finite concatenation of strings which match `P`
+* `P + Q` (`plus P Q`) matches anything which match `P` or `Q`
+* `P * Q` (`comp P Q`) matches `x ++ y` if `x` matches `P` and `y` matches `Q`
+-/
+inductive regular_expression (α : Type u) : Type u
+| zero : regular_expression
+| epsilon : regular_expression
+| char : α → regular_expression
+| plus : regular_expression → regular_expression → regular_expression
+| comp : regular_expression → regular_expression → regular_expression
+| star : regular_expression → regular_expression
+
+namespace regular_expression
+
+instance : inhabited (regular_expression α) := ⟨zero⟩
+
+instance : has_add (regular_expression α) := ⟨plus⟩
+instance : has_mul (regular_expression α) := ⟨comp⟩
+instance : has_one (regular_expression α) := ⟨epsilon⟩
+instance : has_zero (regular_expression α) := ⟨zero⟩
+
+@[simp] lemma zero_def : (zero : regular_expression α) = 0 := rfl
+@[simp] lemma one_def : (epsilon : regular_expression α) = 1 := rfl
+
+@[simp] lemma plus_def (P Q : regular_expression α) : plus P Q = P + Q := rfl
+@[simp] lemma comp_def (P Q : regular_expression α) : comp P Q = P * Q := rfl
+
+/-- `matches P` provides a language which contains all strings that `P` matches -/
+def matches : regular_expression α → language α
+| 0 := 0
+| 1 := 1
+| (char a) := {[a]}
+| (P + Q) := P.matches + Q.matches
+| (comp P Q) := P.matches * Q.matches
+| (star P) := P.matches.star
+
+@[simp] lemma matches_zero_def : (0 : regular_expression α).matches = 0 := rfl
+@[simp] lemma matches_epsilon_def : (1 : regular_expression α).matches = 1 := rfl
+@[simp] lemma matches_add_def (P Q : regular_expression α) :
+  (P + Q).matches = P.matches + Q.matches := rfl
+@[simp] lemma matches_mul_def (P Q : regular_expression α) :
+  (P * Q).matches = P.matches * Q.matches := rfl
+@[simp] lemma matches_star_def (P : regular_expression α) : P.star.matches = P.matches.star := rfl
```

Comment:
If you make this a `ring_hom` you can eliminate some of these lemmas:
```suggestion
def matches_aux : regular_expression α → language α
| 0 := 0
| 1 := 1
| (char a) := {[a]}
| (P + Q) := P.matches + Q.matches
| (comp P Q) := P.matches * Q.matches
| (star P) := P.matches.star

def matches :  regular_expression α →+* language α :=
{ to_fun := matches_aux,
  map_add' := rfl,
  map_mul' := rfl,
  map_zero' := rfl,
  map_one' := rfl }

@[simp] lemma matches.map_star (P : regular_expression α) : P.star.matches = P.matches.star := rfl
```
Let's see if other reviewers think this is a good idea

Reply:
If you like both rings and formal languages, then I should point you to Chapter 9 of the Handbook of Formal Languages, titled "Semirings and Formal Power Series".  A.k.a., regular languages are just solutions to certain linear equation systems in a certain ω-continuous semiring.

Reply:
I don't think I can define this, in this definition `regular_expression` is not a semiring, `P + (Q + R)` and `(P + Q) + R` are not equal as `regular_expression`.

Reply:
Ah, good point